pkg/jsonpath/lexer.go

```diff
@@ -0,0 +1,204 @@
+/*
+Copyright 2023 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jsonpath
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/dop251/goja"
+	js_ast "github.com/dop251/goja/ast"
+)
+
+// Query represents a parsed query in a very restricted subset of JSONPath (RFC 9535).
+// The only supported segment functionality is selecting one definite member of a JSON object.
+type Query []string
+
+// ParseQuery parses a very restricted form of JSONPath expression into a Query.
+// This parser only accepts segments of the form `.name` or `[string]`.
+// A string must be enclosed with double-quotes (`"`).
+// No whitespace is allowed.
+func ParseQuery(queryS string) (Query, error) {
+	lexer, err := NewLexer(queryS)
+	if err != nil {
+		return nil, err
+	}
+	query, err := lexer.ParseQuery()
+	if err != nil {
+		return query, err
+	}
+	if !lexer.eof {
+		return query, fmt.Errorf("syntax error at %d: extra junk %q after last segment", lexer.chrPos, lexer.chr)
+	}
+	return query, nil
+}
+
+type Lexer struct {
+	source string
+	reader io.RuneReader
+
+	chr     rune // next rune to process
+	chrPos  int  // index of start of chr
+	nextPos int  // index after chr
+	eof     bool // no more to process
+}
+
+func NewLexer(source string) (*Lexer, error) {
+	lxr := &Lexer{
+		source: source,
+		reader: strings.NewReader(source),
+	}
+	err := lxr.advance()
+	return lxr, err
+}
+
+func (lxr *Lexer) ParseQuery() (Query, error) {
+	query := Query{}
+	if lxr.chr != '$' {
+		return query, fmt.Errorf("syntax error at %d: missing root identifier (dollar sign)", lxr.chrPos)
+	}
```

Comment:
It does not  seem to handle queries  with an empty string).  Should probably check for that.

Reply:
The empty string is not a valid JSONPath query, and the code will reject that.

Reply:
The latest revision includes a test for this case.